specs/interop/liquidity_migration.md

```diff
@@ -0,0 +1,358 @@
+# Liquidity Migration
+
+<!-- START doctoc generated TOC please keep comment here to allow auto update -->
+<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->
+
+- [Overview](#overview)
+- [Implementation changes](#implementation-changes)
+  - [External functions](#external-functions)
+    - [`convertToSuper`](#converttosuper)
+    - [`convertFromSuper`](#convertfromsuper)
+  - [Events](#events)
+    - [`ConvertedToSuper`](#convertedtosuper)
+    - [`ConvertedFromSuper`](#convertedfromsuper)
+  - [Internal Functions](#internal-functions)
+    - [`checkPair`](#checkpair)
+- [Diagram](#diagram)
+- [Invariants](#invariants)
+- [Considerations](#considerations)
+- [Appendix](#appendix)
+  - [Access Control](#access-control)
+  - [Decimal management](#decimal-management)
+
+<!-- END doctoc generated TOC please keep comment here to allow auto update -->
+
+## Overview
+
+The `OptimismMintableERC20` tokens and `L2StandardToken`s (_legacy tokens_),
+which correspond to locked liquidity in L1, are incompatible with interop.
+Legacy token owners must convert into a `SuperchainERC20` representation, as defined in the [standard](token-bridging.md),
+to move across the Superchain.
+
+The conversion method will use the `L2StandardBridge` mint/burn rights
+over the legacy tokens to allow easy migration to and from the
+corresponding `SuperchainERC20`.
+
+## Implementation changes
+
+The `L2StandardBridge` will be extended to include the following functions and events:
+
+### External functions
+
+#### `convertToSuper`
+
+Converts `_amount` of legacy tokens with address `_legacyAddr`
+to the same `_amount` of `SuperchainERC20` tokens in address `_superAddr`.
+
+```solidity
+convertToSuper(address _legacyAddr, address _superAddr, uint256 _amount)
+```
+
+The function will
+
+1. Check that `_legacyAddr` and `_superAddr` are valid and paired.
+2. Check both tokens have the same amount of decimals.
+3. Burn `_amount` of `_legacyAddr` from `msg.sender`.
+4. Mint `_amount` of `_superAddr` to `msg.sender`.
+
+An example implementation that depends on the `checkPair()`
+access control function looks like this:
+
```

Comment:
Wondering how you feel about the following ABI:

```solidity
function convert(address _from, address _to, uint256 _amount) public {
  require(checkPair(_from, _to));

  if (_isOptimismMintableERC20(_from)) {
    require(_isOptimismSuperchainERC20(_to));
    _mintAndBurn(_from, _to, _amount);
  }
  if (isOptimismMintableERC20(_to)) {
    require(_isOptimismSuperchainERC20(_from));
   _mintAndBurn(_to, _from, _amount);
  }
}

function _mintAndBurn(address _from, address _to, _amount) internal {
    IERC20(_from).burn(msg.sender, _amount);
    IERC20(_to).mint(msg.sender, _amount);
}
```

I think this simplifies the devex as you only have a single function that can be used to convert either direction. The [isOptimismMintableERC20Token](https://github.com/ethereum-optimism/optimism/blob/3c1d3b87573bc3c50eb88a64fff7cd633270d4f4/packages/contracts-bedrock/src/universal/StandardBridge.sol#L398) implementation already exists, we would just need to add one for the superchain erc20

I also think that the decimals check could go inside of `checkPair`

Reply:
The safety of this is in the burn, see [here](https://github.com/Vectorized/solady/blob/1f43cc8005cc3b3c8361dd7dbdd2cdeaf0f99e66/src/tokens/ERC20.sol#L442)

Reply:
Just walking through a user flow here when interacting with a smart contract, you could `approve` a contract, the contract could `transferFrom` and then `convert` and send cross chain

Reply:
@tynes The approach you mention is great IMO. 

This is a timely comment as well, I was debating whether the SuperchainERC20 implementation for convert should inherit from `OptimismMintableERC20` as they share most of the functionality (gives permissions to the bridge, and has a remoteToken it wraps, so very similar to the non-legacy interface). The only thing that annoyed me a bit was inheriting the legacy getters. Do you have any preference on this? As in whether to:

1. Inherit from OptimismMintableERC20 and override `supportsInterface` with `type(IOptimismSuperchainERC20).interfaceId`

2. Don't inherit but copy the non-legacy logic. 

Something to note is that if we want to allow for direct deposits, we should add this new interface to the `StandardBridge` as well

Reply:
We introduced two functions instead of one to keep functions as specific as possible and reduce gas while leaving "ABI" complexities to the integrations. 

That said, the simplification looks elegant, and the gas increase is negligible, so it is probably a good idea.

What do you think @tynes about moving the interface checker inside `checkPair` and move away the conditionals from the main function?

Reply:
Also, `_mintAndBurn()` should only be called as `_mintAndBurn(_from, _to, _amount)`, never as `_mintAndBurn(_to, _from, _amount)`, right?

Reply:
What do you think about something like this?

```solidity
function convert(address _from, address _to, uint256 _amount) public {
  _validatePair(_from, _to);
  IERC20(_from).burn(msg.sender, _amount);
  IERC20(_to).mint(msg.sender, _amount);
  emit Converted(_from, _to, msg.sender, _amount);
}

function _validatePair(address _from, address _to) internal {
  require(IERC20Metadata(_from).decimals() == IERC20Metadata(_to).decimals(), "Decimals do not match")
  if (_isOptimismMintableERC20(_from)) {
    _validateFactories(_from, _to)
  } else {
    _validateFactories(_to, _from)
  }
}

function _validateFactories(address _legacyAddr, address _superAddr) internal {
  address _legacyRemoteToken = factory.deployments(_legacyAddr);
  require(_legacyRemoteToken != address(0), "Invalid Legacy address");

  address _superRemoteToken = superFactory.deployments(_superAddr);
  require(_superRemoteToken != address(0), "Invalid SuperchainERC20 address");
  
  require(_legacyRemoteToken == _superRemoteToken, "Invalid token pair");
}
```

We feel like we can merge the `_isOptimismMintableERC20(...)` check with the `factory.deployments(_from)!=address(0)`, but we should check.

Reply:
This definitely looks good to me. I think its sufficient - no need for the interface checks this way

Reply:
> This is a timely comment as well, I was debating whether the SuperchainERC20 implementation for convert should inherit from OptimismMintableERC20

@0xng I lean towards not inheriting from it given its based on oz rather than solady, we could update the version of it to be based on solady but i think it makes sense to just implement a net new contract. open to ideas tho